lib/srv/db/mysql/engine.go

```diff
@@ -525,3 +535,38 @@ const (
 	// gcpSQLProxyListenPort is the port used by Cloud Proxy for MySQL instances.
 	gcpSQLProxyListenPort = "3307"
 )
+
+// ResolverClients are API clients needed to resolve MySQL endpoints.
+type ResolverClients interface {
+	// GetGCPSQLAdminClient returns GCP Cloud SQL Admin client.
+	GetGCPSQLAdminClient(context.Context) (gcp.SQLAdminClient, error)
+}
+
+// NewEndpointsResolverFunc returns a health check target endpoint resolver.
+func NewEndpointsResolverFunc(ctx context.Context, db types.Database, clients ResolverClients) healthcheck.EndpointsResolverFunc {
+	switch {
+	case db.IsCloudSQL():
+		return newCloudSQLEndpointResolver(db, clients)
+	default:
+		return func(ctx context.Context) ([]string, error) {
+			return []string{db.GetURI()}, nil
+		}
+	}
+}
+
+func newCloudSQLEndpointResolver(db types.Database, clients ResolverClients) healthcheck.EndpointsResolverFunc {
+	return func(ctx context.Context) ([]string, error) {
+		clt, err := clients.GetGCPSQLAdminClient(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		requireSSL, err := cloud.GetGCPRequireSSL(ctx, db, clt)
```

Comment:
this is quite expensive per health check. any better way to handle this? maybe moving it to meta service (which is guranteed to be called when startDatabase)? if it's a lot of work, we could mark it unsupported now and add it separately.